core/http-auth/src/main/java/software/amazon/awssdk/http/auth/internal/DefaultAwsV4HttpSigner.java

```diff
@@ -72,157 +75,225 @@ public SyncSignedRequest sign(SyncSignRequest<? extends AwsCredentialsIdentity>
+                .payload(request.payload().orElse(null))
+                .build();
         }
 
-        SdkChecksum sdkChecksum = createSdkChecksumFromRequest(request.request(), checksumHeaderName, checksumAlgorithm);
-        String contentHash = calculateContentHash(request.payload().orElse(null), sdkChecksum);
+        CompletableFuture<SdkHttpRequest> signedReqFuture =
+            createChecksum(request).thenApply(
+                contentChecksum -> {
+                    SdkHttpRequest.Builder builder = sign(request, contentChecksum);
+                    return builder.build();
+                });
+
 
-        return doSign(request, new ContentChecksum(contentHash, sdkChecksum));
+        return AsyncSignedRequest.builder()
+            .request(CompletableFutureUtils.joinLikeSync(signedReqFuture))
+            .payload(request.payload().orElse(null))
+            .build();
     }
 
-    private SdkHttpRequest.Builder doSign(SignRequest<?, ? extends AwsCredentialsIdentity> request,
+    /**
+     * Using a {@link SignRequest} and a {@link ContentChecksum}, perform all the necessary steps to
+     * create a signed request according to the SigV4 signing documentation:
+     * <p>
+     * https://docs.aws.amazon.com/IAM/latest/UserGuide/create-signed-request.html
+     */
+    protected SdkHttpRequest.Builder sign(SignRequest<?, ? extends AwsCredentialsIdentity> signRequest,
                                           ContentChecksum contentChecksum) {
-        SdkHttpRequest.Builder requestBuilder = request.request().toBuilder();
 
-        Boolean doubleUrlEncode = validatedProperty(request, DOUBLE_URL_ENCODE, true);
-        Boolean normalizePath = validatedProperty(request, NORMALIZE_PATH, true);
-        String checksumHeaderName = request.property(CHECKSUM_HEADER_NAME);
-        Instant requestSigningInstant = validatedProperty(request, REQUEST_SIGNING_INSTANT);
-        String formattedRequestSigningDate = formatDateStamp(requestSigningInstant);
-        String formattedRequestSigningDateTime = formatTimestamp(requestSigningInstant);
-        String regionName = validatedProperty(request, REGION_NAME);
-        String serviceSigningName = validatedProperty(request, SERVICE_SIGNING_NAME);
-        String scope = buildScope(formattedRequestSigningDate, serviceSigningName, regionName);
-
-        AwsCredentialsIdentity sanitizedCredentials = sanitizeCredentials(request.identity());
-        if (sanitizedCredentials instanceof AwsSessionCredentialsIdentity) {
-            addSessionCredentials(requestBuilder, (AwsSessionCredentialsIdentity) sanitizedCredentials);
+        SdkHttpRequest.Builder requestBuilder = signRequest.request().toBuilder();
+        Instant requestSigningInstant = validatedProperty(signRequest, SIGNING_CLOCK).instant();
+        String regionName = validatedProperty(signRequest, REGION_NAME);
+        String serviceSigningName = validatedProperty(signRequest, SERVICE_SIGNING_NAME);
```

Comment:
Was going to initially just comment that validating the requiredness of these this late means it possibly wasted spending cycles in generating contentChecksum, so should do this for optimization/fail-fast. But then realized that with signAsync, the method would return with a CompletableFuture first and fail later if these required properties were missing, which is not the desired behavior. Add input validation to the top of the public methods.

Also, please add tests that validate behavior for missing required properties, for both sign and signAsync.

Reply:
There should probably be some kind of interface method to enforce that. I'll add one.

Reply:
Also, in terms of optimization, we should focus on the most frequent case, because otherwise, if we have to validate properties more than once, we're wasting cycles there. If properties are specified from the caller, it should most always add the correct request parameters, right?

If not, then we should add members to the implementation and set those members when `validatedProperty` is called so that we don't need to get and validate a property more than once.

Reply:
I would agree with adding members, it eases implementation and re-usability; though, at the cost of reason-ability in some ways. This also can be a behavior or the default signers, not the interface, to keep the flexibility.

Reply:
I'm less worried about optimization than the behavior, like the signAsync case I mentioned. So definitely add the unit tests.

Reply:
This is done in the `setParameters()` method.